### ISSUE TITLE
something strange in opera "focus"...

### ISSUE THREAD

**Issue by pyjsorg** [pyjs/original-pyjs#642]
Version
9.50
Build
2042
Platform
Linux
System
i686, 2.6.24-19-generic
Qt library
3.3.8b
Java
Java Runtime Environment installed

JavaScript - http://lkcl.net/jsonrpc/output/
Event thread: click
Error:
name: TypeError
message: Statement on line 12931: Cannot convert undefined or null to Object
Backtrace:
  Line 12931 of inline#2 script in
http://lkcl.net/jsonrpc/output/Opera.cache.html
            elem.firstChild.focus();
  ...  Line 12935 of inline#2 script in
http://lkcl.net/jsonrpc/output/Opera.cache.html
            return
**ui_Focus.prototype.focus.call.apply(__ui_Focus.prototype.focus, arguments);
  Line 4769 of inline#2 script in
http://lkcl.net/jsonrpc/output/Opera.cache.html
        __ui_Focus.prototype.__class**.focus(this, this.getElement());
  Line 14913 of inline#2 script in
http://lkcl.net/jsonrpc/output/Opera.cache.html
        this.subject.setFocus(true);
  ...  Line 14859 of inline#2 script in
http://lkcl.net/jsonrpc/output/Opera.cache.html: In function SendMail_MailPopup
        if(instance.**init**) instance.**init**.apply(instance, arguments);
  Line 15658 of inline#2 script in
http://lkcl.net/jsonrpc/output/Opera.cache.html
        var popup = SendMail_MailPopup('message', 'Send Message',
this.sink, this.user);
  Line 4701 of inline#2 script in
http://lkcl.net/jsonrpc/output/Opera.cache.html
        listener.onClick(this, event);
  Line 3476 of inline#2 script in
http://lkcl.net/jsonrpc/output/Opera.cache.html: In function
DOM_dispatchEventImpl
        listener.onBrowserEvent(event);
  Line 3448 of inline#2 script in
http://lkcl.net/jsonrpc/output/Opera.cache.html: In function DOM_dispatchEvent
        DOM_dispatchEventImpl(event, element, listener);
  Line 2620 of inline#2 script in
http://lkcl.net/jsonrpc/output/Opera.cache.html
                DOM_dispatchEvent(evt, curElem, listener);
  ...
stacktrace: n/a; see 'opera:config#UserPrefs|Exceptions Have Stacktrace'

_Original issue: http://code.google.com/p/pyjamas/issues/detail?id=74 (June 27, 2008 15:10:56)_

**Comment by pyjsorg** [comment]
_From g.clinch@gmail.com on November 27, 2008 23:46:19:_
I get what looks like the same error with Safari (3.2 on OS X) - 'Null value' at the elem.firstChild.focus() call in:

   __ui_Focus.prototype.focus = function(elem) {

```
    elem.firstChild.focus();

};
```

Equally, google's Chrome (0.4.154.25) which also uses webkit gives:
Uncaught TypeError: Cannot call method 'focus' of null.

**Comment by pyjsorg** [comment]
_From g.clinch@gmail.com on December 03, 2008 22:16:03:_
I seem to have fixed this in my local checkout for both Safari and Opera, but I wonder if actually what is 
needed is more use of createFocusable(), as I think my change will have broken Trees and FocusPanels - 
neither of which I'm currently using.

My changes:

# Index: library/platform/uiSafari.py

--- library/platform/uiSafari.py        (revision 161)
+++ library/platform/uiSafari.py        (working copy)
@@ -3,7 +3,7 @@

```
 def blur(self, elem):
     JS("""
```
-        elem.firstChild.blur();
- ```
     elem.blur();
   """)
  ```
  
   def createFocusable(self):
  @@ -36,7 +36,7 @@
  
   def focus(self, elem):
       JS("""
-        elem.firstChild.focus();
- ```
     elem.focus();
   """)
  ```
  
   def getTabIndex(self, elem):
  
  # Index: library/platform/uiOpera.py
  
  --- library/platform/uiOpera.py (revision 161)
  +++ library/platform/uiOpera.py (working copy)
  @@ -3,7 +3,7 @@
  
   def blur(self, elem):
       JS("""
-        elem.firstChild.blur();
- ```
     elem.blur();
   """)
  ```
  
   def createFocusable(self):
  @@ -36,7 +36,7 @@
  
   def focus(self, elem):
       JS("""
-        elem.firstChild.focus();
- ```
     elem.focus();
   """)
  ```
  
   def getTabIndex(self, elem):

**Comment by pyjsorg** [comment]
_From luke.lei...@gmail.com on December 04, 2008 09:05:39:_
done for opera.

i took a closer look at gwt 1.5, they use a timeout of 0 to get the function to be
called from _outside_ the event handler, can you let me know if it works?

svn update.

**Comment by pyjsorg** [comment]
_From g.clinch@gmail.com on December 04, 2008 23:15:45:_
Can confirm it's fixed for Opera (9.62 on XP).

Still broken on Safari (Error console now gives just 'Null value' with no line
indication), and on Google Chrome (Uncaught TypeError: Cannot call method 'focus' of
null).

(This is trying to setFocus(True) on a TextBox, if it helps!)

The following fixes it for me:

# Index: library/platform/uiSafari.py

--- library/platform/uiSafari.py        (revision 163)
+++ library/platform/uiSafari.py        (working copy)
@@ -7,7 +7,7 @@
         // generally be unsuccessful, so we invoke blur() from outside of
         // the callback.
         $wnd.setTimeout(function() {
-                                       elem.firstChild.blur();
-                                       elem.blur();
                       },
                       0);
       """)
  @@ -46,7 +46,7 @@
       // generally be unsuccessful, so we invoke focus() from outside of
       // the callback.
       $wnd.setTimeout(function() {
-                                       elem.firstChild.focus();
-                                       elem.focus();
                       },
                       0);
       """)

**Comment by pyjsorg** [comment]
_From luke.lei...@gmail.com on December 05, 2008 09:21:47:_
ok - well, there we go.  strange but true: gwt 1.5 appears broken...
